### ISSUE TITLE
Function generics parameters not inferring correctly when programmatically generated

### ISSUE THREAD

**Issue by ecyrbe** [microsoft/TypeScript#57049]
### 🔎 Search Terms

I searched all bugs related to functions generics and did not found this bug.

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about  generic functions

### ⏯ Playground Link

https://tsplay.dev/mALzPN

### 💻 Code

```ts
import { TypeSkell } from 'typeskell';
import { Kind } from 'typeskell/dist/kinds';
import * as R from 'typeskell/dist/data/result';
import { pipe } from 'typeskell/dist/pipe';

interface Functor<F extends Kind> {
    // this is using a compile time utility type doing AST transformations internally to generate a generic function depending on F HKT arity
    map: TypeSkell<'(a -> b) -> F a ..e -> F b ..e', { F: F }>;
};

interface Functor2<F extends Kind> {
    map: <A1, B1>(f: (a: A1) => B1) => TypeSkell<'F a ..e -> F b ..e', { F: F }, { a: A1, b: B1 }>;
};

const RImpl: Functor<R.TResult> = {
    map: f => fa => R.isOk(fa) ? R.ok(f(fa.ok)) : fa
}

const R2Impl: Functor2<R.TResult> = {
    map: f => fa => R.isOk(fa) ? R.ok(f(fa.ok)) : fa
}


type MapImpl = typeof RImpl.map;
//   ^? type MapImpl = <A1, A2>(args_0: (args_0: A1) => A2) => <B1>(args_0: R.Result<A1, B1>) => R.Result<A2, B1>

type Map2Impl = typeof R2Impl.map;
//   ^? type Map2Impl = <A1, B1>(f: (a: A1) => B1) => <A1>(args_0: R.Result<A1, A1>) => R.Result<B1, A1>
//   same signature a MapImpl

type Map = typeof R.map;
//   ^? type Map = <A1, A2>(args_0: (args_0: A1) => A2) => <B1>(args_0: R.Result<A1, B1>) => R.Result<A2, B1>
//   same signature a MapImpl

const testImpl = pipe(R.ok(0), RImpl.map(a => a + 1));
//     ^? const testImpl: R.Result<unknown, never>
// Result is badly inferred as Result<unknown, never>

const testImpl2 = pipe(R.ok(0), R2Impl.map(a => a + 1));
//     ^? const testImpl2: R.Result<number, never>
// Result is correctly inferred as Result<number, never>

const test = pipe(R.ok(0), R.map(a => a + 1));
//     ^? const test: R.Result<number, never>
// Result is correctly inferred as Result<number,never>
```

### 🙁 Actual behavior

Signature is ok but when calling `map`, type get badly inferred
However, when compiling the Functor on a resolveld HKT (here a result), type Result gets correctly inferred

### 🙂 Expected behavior

I expect that `map` infers correctly the Result type as if i wrote the signature by hand and not programmatically

### Additional information about the issue

This is using an experimental library i developped to improve HKT libraries implementations like fp-ts.
Indeed, fp-ts current implementation uses a lot function overloading, but the solution does not scale, you need to implement each Typeclass for all possible arity of your Typeconstructor.
This compile time magic allows to implement only one signature one time.
However, i'm stuck on typescript not inferring the types correctly even if the generated generic function signature is correct.

Link to the library for further investigation : https://github.com/ecyrbe/typeskell

**Comment by craigphicks** [comment]
It's hard to get a handle of this because the description is not self contained in the post.  If you could describe the problem without reference to any external libraries that would be a start.  The tiniest case that presents your bug is the most likely to either get the bug fixed, or let you know about the limitations.

A separate comment that doesn't exactly belong here - I looked at your link *"Knowning variance allows to know if we should & or | the generic types programatically and also know what are the default values (unknown for invariants and contravariants, and never for covariants)."*  Great stuff!  It sounds very interesting.  If there were some tiny and easy examples of the programatically generated types in the README it would be helpful.

**Comment by RyanCavanaugh** [comment]
+1 to the above comment; we need an isolated repro to have a chance at diagnosing this.

**Comment by ecyrbe** [comment]
I have found a workaround thanks to the help of @Andarist .
The thing is that when programmatically building functions and genericfunctions, if you pass a type to a custom utility type, it get fixed by typescript even if it's not used in the output of the utility type. So typescript reads the params and fixes it, and since it can't resolve it (since it not used in the output) it gets inferred as unknown.
The workaround consists of filtering the type parameters passed to the utility types to only have the ones that are actually referenced.
Closing
